### PR TITLE
feat(service): implement service class

### PR DESCRIPTION
**Pull request by khjobran1**
Mostly Done.... 🚧

### CODE REVIEW

File: lib/msd_odata/url_helper.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
+module MsdOdata
+  class UrlHelper
+    def initialize(base_url, entity)
+      @base_url = base_url
+      @entity = entity
+    end
+
+    def entity_collection_url
+      "#{@base_url}/data/#{@entity.name}?#{@entity.query}"
```
Comment (hamdyfci) [inline lib/msd_odata/url_helper.rb:9]: what is `@entity.query`?

File: lib/msd_odata/url_helper.rb [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    end
+
+    def entity_url
+      raise "'query_params' are missing." unless @entity.attrs[:query_params].is_a? Hash
+      params = @entity.attrs[:query_params].map { |k,v| "#{k}='#{v}'"}.join(',')
+      "#{entity_collection_url}(#{params})?#{@entity.query}"
```
Comment (hamdyfci) [inline lib/msd_odata/url_helper.rb:15]: also here what is `@entity.query`?

File: lib/msd_odata/url_helper.rb (hunk 1 above)
Comment (khjobran1) [reply lib/msd_odata/url_helper.rb:9]: I forgot to remove it, still working on it.

Should I remove it now so we can merge the PR? or should I complete everything in this PR then we merge?

File: lib/msd_odata/service.rb [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    def initialize(token, base_url, entity)
+      @entity = entity
+      @url_helper = UrlHelper.new(base_url, entity)
+      @options = {
+        headers: { 'Authorization': token },
+      }
+    end
+
+    def create
+      @options[:body] = @entity.attrs[:body]
```
Comment (hamdyfci) [inline lib/msd_odata/service.rb:12]: prefer in this case to use only `@entity.attrs`. i.e:
when the caller tries to create the `attrs` will be the attributes of a valid entity.
and when trying to `read, update, or delete` will be two parts ie: 

```ruby
atttrs = { query: {}, body: {}}
```

Comment (khjobran1) [reply lib/msd_odata/service.rb:12]: I thought it will be consistent to have the same structure in all actions

**Comment by khjobran1** [comment]
> also, it will be great if you can add some tests

Sure

File: lib/msd_odata/url_helper.rb (hunk 1 above)
Comment (hamdyfci) [reply lib/msd_odata/url_helper.rb:9]: I think we can't merge this PR before @EslamM-H's one( that contains the `Client` class).
So I prefer to complete everything in this PR

File: lib/msd_odata/service.rb (hunk 3 above)
Comment (hamdyfci) [reply lib/msd_odata/service.rb:12]: I think they are two different things, as I'm thinking of it as an active record model when the caller creates he is expecting to pass the attributes to be created.

Also, I think it should be added to the readme file explaining the usage

Comment (khjobran1) [reply lib/msd_odata/service.rb:12]: Yes its easier to have it as one hash for create. Will do.

File: spec/url_helper_spec.rb [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+      entity = MsdOdata::Entity.new('Customers', { url_params: { 'CustomerId' => 'HS-120' } })
+      base_url = 'github.com'
+      url_helper = MsdOdata::UrlHelper.new(base_url, entity)
+
+      expected_url = "github.com/data/Customers(CustomerId='HS-120')"
+
+      expect(url_helper.entity_url).to eq(expected_url)
+    end
+  end
+end
```
Comment (EslamM-H) [inline spec/url_helper_spec.rb:27]: This file should be moved from `spec/url_helper_spec.rb` to be `spec/lib/msd_odata/url_helper_spec.rb` or to `spec/lib/msd_odata/util/url_helper_spec.rb` if you choose to move `url_helper.rb` to `lib/msd_odata/util`

File: lib/msd_odata/url_helper.rb [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+      query += "$skip=#{query_hash[:skip]}&" unless query_hash[:skip].nil?
+      query
+    end
+
+    # Parses 'url_params' hash to look like: "key='value',key='value'"
+    def build_url_params
+      @entity.attrs[:url_params].map { |k, v| "#{k}='#{v}'" }.join(',')
+    end
+  end
+end
```
Comment (EslamM-H) [inline lib/msd_odata/url_helper.rb:41]: I think it would be better to move this class to `util`

File: spec/service_spec.rb [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+      expected_options = {
+        headers: { 'Authorization': 'some_token' },
+      }
+
+      expect_any_instance_of(MsdOdata::Client).to receive(:request).with(:get, expected_options)
+
+      service.find_entity
+    end
+  end
+end
```
Comment (EslamM-H) [inline spec/service_spec.rb:96]: This file should be moved from `spec/service_spec.rb` to be `spec/lib/msd_odata/service_spec.rb`

Comment (khjobran1) [reply spec/service_spec.rb:96]: Yeah I'm waiting for you PR to be merged and I will resolve the conflicts and arrange the files

File: lib/msd_odata/service.rb [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+      url = @url_helper.entity_url
+      request(:delete, url)
+    end
+
+    ##
+    # Finds an entity of a resource
+    # GET request to 'base_url/data/EntityNames(url_params)'
+    #
+    # @return [response]
+    def find_entity
```
Comment (hamdyfci) [inline lib/msd_odata/service.rb:60]: I would prefer to name it just `find`

Comment (khjobran1) [reply lib/msd_odata/service.rb:60]: OK

File: lib/msd_odata/service.rb [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+      request(:post, url)
+    end
+
+    ##
+    # Queries a resource
+    # GET request to 'base_url/data/EntityNames'
+    #
+    # @return [response]
+    def read
+      url = @url_helper.entity_collection_url
```
Comment (EslamM-H) [inline lib/msd_odata/service.rb:29]: I think this needs to be `url = @url_helper.entity_collection_url(with_query: true)` to build the query in read action

Comment (khjobran1) [reply lib/msd_odata/service.rb]: Done

**Pull request merged**